commitrmnocb/outcome.go

```diff
@@ -0,0 +1,316 @@
+// were observed when the most recent report was generated. If an update to these max seq sums is detected, it means
+// that the previous report has been transmitted, and we output ReportTransmitted to dictate that a new report
+// generation phase should begin. If no update is detected, and we've exhausted our check attempts, output
+// ReportTransmissionFailed to signify we stop checking for updates and start a new report generation phase. If no
+// update is detected, and we haven't exhausted our check attempts, output ReportInFlight to signify that we should
+// check again next round.
+func (p *Plugin) checkForReportTransmission(
+	previousOutcome Outcome,
+	consensusObservation ConsensusObservation,
+) Outcome {
+
+	offRampUpdated := false
+	for _, previousSeqNumChain := range previousOutcome.OffRampNextSeqNums {
+		if currentSeqNum, exists := consensusObservation.OffRampNextSeqNums[previousSeqNumChain.ChainSel]; exists {
+			if previousSeqNumChain.SeqNum != currentSeqNum {
+				offRampUpdated = true
+				break
+			}
+		}
+	}
+
+	if offRampUpdated {
+		return Outcome{
+			OutcomeType: ReportTransmitted,
+		}
+	}
+
+	if previousOutcome.ReportTransmissionCheckAttempts+1 >= p.cfg.MaxReportTransmissionCheckAttempts {
+		p.lggr.Warnw("Failed to detect report transmission")
+		return Outcome{
+			OutcomeType: ReportTransmissionFailed,
+		}
+	}
+
+	return Outcome{
+		OutcomeType:                     ReportInFlight,
+		OffRampNextSeqNums:              previousOutcome.OffRampNextSeqNums,
+		ReportTransmissionCheckAttempts: previousOutcome.ReportTransmissionCheckAttempts + 1,
+	}
+}
+
+// getConsensusObservation Combine the list of observations into a single consensus observation
+func (p *Plugin) getConsensusObservation(aos []types.AttributedObservation) (ConsensusObservation, error) {
+	aggObs := aggregateObservations(aos)
+	fChains := p.fChainConsensus(aggObs.FChain)
+
+	fDestChain, exists := fChains[p.cfg.DestChain]
+	if !exists {
+		return ConsensusObservation{},
+			fmt.Errorf("no consensus value for fDestChain, DestChain: %d", p.cfg.DestChain)
+	}
+
+	consensusObs := ConsensusObservation{
+		MerkleRoots: p.merkleRootConsensus(aggObs.MerkleRoots, fChains),
+		// TODO: use consensus of observed gas prices
+		GasPrices: make(map[cciptypes.ChainSelector]cciptypes.BigInt),
+		// TODO: use consensus of observed token prices
+		TokenPrices:        make(map[types.Account]cciptypes.BigInt),
+		OnRampMaxSeqNums:   p.onRampMaxSeqNumsConsensus(aggObs.OnRampMaxSeqNums, fChains),
+		OffRampNextSeqNums: p.offRampMaxSeqNumsConsensus(aggObs.OffRampNextSeqNums, fDestChain),
+		FChain:             fChains,
+	}
+
+	return consensusObs, nil
+}
+
+// Given a mapping from chains to a list of merkle roots, return a mapping from chains to a single consensus merkle
+// root. The consensus merkle root for a given chain is the merkle root with the most observations that was observed at
+// least fChain times.
+func (p *Plugin) merkleRootConsensus(
```

Comment:
There's a generic helper to manage counting observations that could be put in a reusable location: `execute/internal/validation/reports.go`